File: CSharp/Cs_Generic/Stats.cs
/*-------------------------------------------------------------------
  Stats.cs
  - provides definitions for user-defined class Stats<T>
*/

using System.Collections;
using System.Collections.Generic;   // IEnumerable<T>, List<T>, ...
using Analysis;

namespace Stats {
  /*----------------------------------------------------------------------
    Stats<T> is a generic demo type 
     - holds any finite number of generic values
     - computes several statistics on its values
  */
  public class Stats<T> 
    where T: IComparable, IComparable<T>, IEquatable<T>, IConvertible
  {
    /*--------------------------------------------------------------------
      Constructs a point with N coordinates each with default value
    */
    public List<T> items { get; set; }
    public Stats(List<T> l) {
      items = l;
    }
    public void check() {
      if (items == null) {
        throw new SystemException("stats is null");
      }
      else if (items.Count < 1)
        throw new SystemException("stats is empty");
    }
    public int size() {
      check();
      return items.Count; 
    }
    public T max() {
      check();
      T max = items[0];
      foreach(T item in items) {
        if(max.CompareTo(item) < 0)
          max = item;
      }
      return max;
    }
    public T min() {
      check();
      T min = items[0];
      foreach(T item in items) {
        if(min.CompareTo(item) > 0)
          min = item;
      }
      return min;
    }
    public T? sum() {
      check();
      dynamic? sum = default(T);
      if(sum != null) {
        foreach(T item in items) {
          sum += item;
        }
        return sum;
      }
      return default(T);
    }
    public T avg() {
      check();
      dynamic? sum = this.sum();
      return sum/items.Count;
    }
  }
}

File: Python/Py_Iter/Stats.py
#----------------------------------------------------------
#   Py_Generic::Stats.py
#   Stats[T]
#     Collection of T items with simple statistics.
#----------------------------------------------------------

import AnalysisIter
from typing import TypeVar, Generic, Iterator

# Generic point class with N coordinates
# - Supports indexing and iteration

T = TypeVar('T', bound=int | float)

Anal = AnalysisIter

class Stats(Generic[T]):
  def __init__(self, l:list[T]) -> None:
    self.items = l

  def len(self) -> int:
    return self.items.len()

  def max(self) -> T | None:
    if not self.items:
      return None
    max:T = self.items[0]
    for item in self.items:
      if max < item:
        max = item
    return max
  
  def min(self) -> T | None:
    if not self.items:
      return None
    min:T = self.items[0]
    for item in self.items:
      if min > item:
        min = item
    return min
  
  def sum(self) -> T:
    sum:T = 0
    for item in self.items:
      sum += item
    return sum
  
  def avg(self) -> T | None:
     if not self.items:
       return None
     num = self.sum()
     den = len(self.items)
     return num/den
     
  def __iter__(self) -> Iterator:                 # p.iter()
      return self.items.__iter__()
